calamus/fields.py

```diff
@@ -178,15 +180,63 @@ def _serialize(self, value, attr, obj, **kwargs):
 class DateTime(_JsonLDField, fields.DateTime):
     """A date/time field."""
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, extra_formats=("%Y-%m-%d",), **kwargs):
         super().__init__(*args, **kwargs)
+        self._extra_formats = extra_formats
 
     def _serialize(self, value, attr, obj, **kwargs):
         value = super()._serialize(value, attr, obj, **kwargs)
         if self.parent.opts.add_value_types:
             value = {"@value": value, "@type": "http://www.w3.org/2001/XMLSchema#dateTime"}
         return value
 
+    def _deserialize(self, value, attr, data, **kwargs):
+        try:
+            return super()._deserialize(value, attr, data, **kwargs)
+        except ValidationError:
+            pass
+
+        # Try with extra formats
+        for format in self._extra_formats:
+            try:
+                original_format = self.format
+                self.format = format
+                return super()._deserialize(value, attr, data, **kwargs)
+            except ValidationError:
+                pass
+            finally:
+                self.format = original_format
+
+        raise self.make_error("invalid", input=value, obj_type=self.OBJ_TYPE)
+
+
+class Uri(_JsonLDField, fields.String, fields.Dict):
```

Comment:
That we have something that can be either a `string` or a `class Url` is a very renku specific thing and not really something we want to have in calamus.

I think in these cases, the user would either have to add a custom field type on their end (in renku-python in this case) or we should add a `MultiType` or `VariableType` field where a user specifies a selector function/strategy along with types it applies to, so we can solve the problem in a generic way.

Reply:
Moved to renku.